text/2442-simple-postfix-macros.md

```diff
@@ -0,0 +1,351 @@
+extensions; it's not possible to experiment with postfix syntax in macro-based
+extensions to the language.  Language changes often start through
+experimentation, and such experimentation can also result in sufficiently good
+alternatives to avoid requiring language extensions. Having a postfix macro
+syntax would serve both goals: enabling the prototyping of new language
+features, and providing compelling syntax without having to extend the
+language.
+
+Previous discussions of method-like macros have stalled in the process of
+attempting to combine properties of macros (such as unevaluated arguments) with
+properties of methods (such as type-based or trait-based dispatch). This RFC
+proposes a minimal change to the macro system that allows defining a simple
+style of postfix macro, inspired specifically by `try!(expr)` becoming `expr?`
+and `await!(expr)` becoming `expr.await`, without blocking potential future
+extensions. In particular, this RFC does not in any way preclude a future
+implementation of postfix macros with full type-based or trait-based dispatch.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a macro using `macro_rules!`, you can include a first argument
+that uses a designator of `self` (typically `$self:self`). This must appear as
+the first argument, and outside any repetition. If the macro includes such a
+case, then Rust code may invoke that macro using the method-like syntax
+`expr.macro!(args)`. The Rust compiler will expand the macro into code that
+receives the evaluated `expr` as its first argument.
+
+```rust
+macro_rules! log_value {
+    ($self:self, $msg:expr) => ({
+        eprintln!("{}:{}: {}: {:?}", file!(), line!(), $msg, $self);
+        $self
+    })
+}
+
+fn value<T: std::fmt::Debug>(x: T) -> T {
+    println!("evaluated {:?}", x);
+    x
+}
+
+fn main() {
+    value("hello").log_value!("value").len().log_value!("len");
+}
+```
+
+This will print:
+
+```
+evaluated "hello"
+src/main.rs:14: value: "hello"
+src/main.rs:14: len: 5
+```
+
+Notice that `"hello"` only gets evaluated once, rather than once per reference
+to `$self`, and that the `file!` and `line!` macros refer to the locations of
+the invocations of `log_value!`.
+
+A macro that accepts multiple combinations of arguments may accept `$self` in
+some variations and not in others. For instance, a macro `do_thing` could allow
+both of the following:
+
+```rust
+do_thing!(some_expression());
+some_other_expression().do_thing!().further_expression().do_thing!();
+```
+
+This method-like syntax allows macros to cleanly integrate in a left-to-right
+method chain, while still making use of control flow and other features that
+only a macro can provide.
+
+A postfix macro may accept `self` by value, by reference, or by mutable
+reference; the compiler will automatically use the appropriate type of
+reference, just as it does for closure captures.
+
+# Reference-level explanation
```

Comment:
Just re-read this RFC with a specific question in mind, and I didn't find an answer:

If this RFC is accepted (and implemented), will it be allowed to define a macro that has one or more `($self, ...)` rules **and** one or more rules without `$self`, i.e. a macro that can be used as both a function-like macro and a method-like macro? This would allow some macros that exist today to add postfix rules backwards-compatibly, and could be useful for some things outside of backwards compat, like my [`as_variant!`](https://docs.rs/as_variant/latest/as_variant/macro.as_variant.html) macro (the first two rules are no longer very interesting with postfix macros, but the other two are).

Reply:
@jplatte Added an explicit note to that effect, thank you. Macros *can* support both, by having some rules with a `self` designator and some rules without.